examples/07_tutorials/ai-reco-of-recos/utils/openai_utils.py

```diff
@@ -0,0 +1,456 @@
+    after_log
+)
+
+from utils.openai_data_models import *
+from utils.file_utils import convert_png_to_jpg, get_image_base64
+
+
+
+def get_encoder(model = "gpt-4o"):
+    if model == "gpt-45":
+        return tiktoken.get_encoding("o200k_base")       
+    if model == "gpt-4o":
+        return tiktoken.get_encoding("o200k_base")
+    if model == "gpt-4.1":
+        return tiktoken.get_encoding("o200k_base")           
+    if model == "o1":
+        return tiktoken.get_encoding("o200k_base")       
+    if model == "o1-mini":
+        return tiktoken.get_encoding("o200k_base")       
+    if model == "o3":
+        return tiktoken.get_encoding("o200k_base")
+    if model == "o3-mini":
+        return tiktoken.get_encoding("o200k_base")
+    if model == "o4-mini":
+        return tiktoken.get_encoding("o200k_base")
+    else:
+        return tiktoken.get_encoding("o200k_base")
+
+
+def get_token_count(text, model = "gpt-4o"):
+    enc = get_encoder(model)
+    return len(enc.encode(text))
+
+
+def prepare_image_messages(imgs):
+    img_arr = imgs if isinstance(imgs, list) else [imgs]
+    img_msgs = []
+
+    for image_path_or_url in img_arr:
+        if image_path_or_url.startswith("http"):
+            image = image_path_or_url
+        else:
+            image_path_or_url = os.path.abspath(image_path_or_url)
+            try:
+                if os.path.splitext(image_path_or_url)[1] == ".png":
+                    image_path_or_url = convert_png_to_jpg(image_path_or_url)
+                image = f"data:image/jpeg;base64,{get_image_base64(image_path_or_url)}"
+            except:
+                image = image_path_or_url
+
+        img_msgs.append({ 
+            "type": "image_url",
+            "image_url": {
+                "url": image
+            }
+        })
+    # console.print("Image messages prepared.")
+    return img_msgs
+
+
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))             
+def get_embeddings(text : str, model_info: EmbeddingModelnfo = EmbeddingModelnfo()):
+    if model_info.client is None: model_info = instantiate_model(model_info)
+    return model_info.client.embeddings.create(input=[text], model=model_info.model_name).data[0].embedding
+
+
+
+def call_llm(prompt: str, model_info: Union[MulitmodalProcessingModelInfo, TextProcessingModelnfo], temperature = 0.2, imgs=[]):
+    content = [{"type": "text", "text": prompt}]
+    content = content + prepare_image_messages(imgs)
+    messages = [
+        {"role": "user", "content": "You are a helpful assistant that processes text and images."},
+        {"role": "user", "content": content},
+    ]
+    
+    if model_info.client is None: model_info = instantiate_model(model_info)
+    # print(">>>>>>>>>>>>>>>>> call_llm model_info", model_info)
+
+    if (model_info.model_name == "gpt-4o") or ((model_info.model_name == "gpt-45")):
+        return call_4(messages, model_info.client, model_info.model, temperature)
```

Comment:
You may simplify this conditional logic by distinguishing between reasoning and non-reasoning models instead of listing all the models individually.

I would recommend using the Responses API instead of the Chat API for this type of task though.